docs: first indexer cheat sheet version (#3433)
jina-ai/jina: docs/advanced/experimental/scaling.md

# Scaling Executors in kubernetes

> In Jina we support to ways of scaling:

Reviewer: ```suggestion
In Jina we support two ways of scaling:
```

- *Replicas* can be used with any Executor type.
- *Shards* should only be used with Indexers, since they store a state.

```{admonition} Important
:class: important

> This page describes Jinas behavior with kubernetes deployment for now.

Reviewer: ```suggestion
This page describes Jina's behavior with kubernetes deployment for now.
```

When using `shards` and `replicas` without kubernetes, Jina will behave slightly different.

> It is discouraged to use the `parallel` argument, when deploying to kubernetes.

Reviewer: ```suggestion
It is discouraged to use the `parallel` argument when deploying to kubernetes.
```

```

(replicas)=
## Replicas

Replication (or horizontal scaling) means duplicating an Executor and its state.
It allows more requests to be served in parallel.
A single request to Jina is only send to one of the replicas.
Furthermore, it increases the failure tolerance.

> When one replica dies, another will immediately take over.

Reviewer: But this is not true right now, is it? Do our replicas restart if they die? Or is this about k8s?

In a cloud environment any machine might die at any point in time.
Thus, using replicas is important for reliable services.

> Replicas are currently only supported, when deploying Jina with kubernetes.

Reviewer: Ah I see. So we're not mentioning the built-in replicas here.

```python Usage
from jina import Flow

f = Flow().add(name='ExecutorWithReplicas', replicas=3)
```

(shards)=
## Shards

Sharding means splitting the content of an Indexer into several parts.
These parts are then put on different machines.

> This is helpful, in two situations:

Reviewer: ```suggestion
This is helpful in two situations:
```

- When the full data does not fit onto one machine.

> - When the latency of a single request becomes to slow.

Reviewer: ```suggestion
- When the latency of a single request becomes too slow.
```

>   Then two machines can compute the result faster.

Reviewer: ```suggestion
  Then splitting the load across two or more machines yields better results.
```

When you shard your index, the request handling usually differes for index and search requests:

- Index (and update, delete) will just be handled by a single shard.
- Search requests are handled by all shards.

```python Usage
from jina import Flow

> f = Flow().add(name='ExecutorWithShards', shards=3)

Reviewer: Maybe mention uses_after, or link to it, if it's discussed somewhere else.

Author: Actually, uses_after is nowhere explained in detail. So here is the right place to do so.

Reviewer: Okay, makes sense

```

## Combining Replicas & Shards

> Combining both gives all above mentioned advantages.

Reviewer: ```suggestion
Combining both gives all the advantages mentioned above.
```

> When deploying to kubernetes, Jina replicates is applied on each single shard.

Reviewer: the second part is confusing. Replicates what?

Author: "When deploying to kubernetes, Jina replicates each single shard." better?

Reviewer: Yep

When one replica dies, another will immediately take over.
In a cloud environment any machine might die at any point in time.
Thus, using replicas is important for reliable services.
Replicas are currently only supported, when deploying Jina with kubernetes.

```python Usage
from jina import Flow

f = Flow().add(name='ExecutorWithReplicas', replicas=3)
```

(shards)=
## Shards

Sharding means splitting the content of an Indexer into several parts.
These parts are then put on different machines.
This is helpful in two situations:

- When the full data does not fit onto one machine.
- When the latency of a single request becomes too slow.
  Then splitting the load across two or more machines yields better results.

When you shard your index, the request handling usually differes for index and search requests:

- Index (and update, delete) will just be handled by a single shard => `polling='any'`
- Search requests are handled by all shards => `polling='all'`

```python Usage
from jina import Flow

index_flow = Flow().add(name='ExecutorWithShards', shards=3, polling='any')
search_flow = Flow().add(name='ExecutorWithShards', shards=3, polling='all', uses_after='MatchMerger')
```

> ### Merging search results via `uses_after`

Reviewer: ```suggestion
### Merging Search Results Via `uses_after`
```
